From: Ira Weiny <ira.weiny@intel.com>
Subject: mm-userfaultfd-replace-kmap-kmap_atomic-with-kmap_local_page-v2
Date: Tue, 25 Oct 2022 15:01:36 -0700

Update the commit message and comment based on additional discussion

Thanks to Matt for pointing out the deadlock potential despite recursive
reads.  

Link: https://lkml.kernel.org/r/20221025220136.2366143-1-ira.weiny@intel.com
Signed-off-by: Ira Weiny <ira.weiny@intel.com>
Cc: Matthew Wilcox <willy@infradead.org>
Cc: Andrew Morton <akpm@linux-foundation.org>
Cc: Andrea Arcangeli <aarcange@redhat.com>
Cc: Peter Xu <peterx@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/userfaultfd.c |   16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

--- a/mm/userfaultfd.c~mm-userfaultfd-replace-kmap-kmap_atomic-with-kmap_local_page-v2
+++ a/mm/userfaultfd.c
@@ -159,9 +159,19 @@ static int mcopy_atomic_pte(struct mm_st
 
 		page_kaddr = kmap_local_page(page);
 		/*
-		 * The mmap_lock is held here.  Disable page faults to
-		 * prevent deadlock should copy_from_user() fault.  The
-		 * copy will be retried outside the mmap_lock.
+		 * The read mmap_lock is held here.  Despite the
+		 * mmap_lock being read recursive a deadlock is still
+		 * possible if a writer has taken a lock.  For example:
+		 *
+		 * process A thread 1 takes read lock on own mmap_lock
+		 * process A thread 2 calls mmap, blocks taking write lock
+		 * process B thread 1 takes page fault, read lock on own mmap lock
+		 * process B thread 2 calls mmap, blocks taking write lock
+		 * process A thread 1 blocks taking read lock on process B
+		 * process B thread 1 blocks taking read lock on process A
+		 *
+		 * Disable page faults to prevent potential deadlock
+		 * and retry the copy outside the mmap_lock.
 		 */
 		pagefault_disable();
 		ret = copy_from_user(page_kaddr,
_
